### ISSUE TITLE
Cool stuff. Feature: enable history to be a trailing contenxt of n messages.

### ISSUE THREAD

**Issue by syonfox** [drorm/gish#1]
First of all nice project, I appreciate a clean cli template for js :)

I think a useful addition would be to allow history.
```
let messages = [] 

onNewMessage(msg) {
  let trash = messages.shift();
  messages.push(msg);
}
```

**Comment by syonfox** [comment]
One other idea is for long messages to get the ai to summarize and then save that as the history

**Comment by drorm** [comment]
History is already saved in a log file: https://github.com/drorm/gish/blob/main/README.md#log-file in JSON format, both the request and the response.
Do you mean, provide a way to retrieve it as in
```
gish -h
```

**Comment by syonfox** [comment]
Sorry for the confusion.  That may be useful but I meant a way to support a trailing history context being passed with the requests to openai. 

the problem is when engaging in an interactive season there is no history or context being fead to the model.  It would be interesting to have a way of defining this.  

For example a way of seating a system message and then also providing a trailing history.

For example options 
```
--system "Please be a helpful ai assistant that is funny and tells bad jokes."
--history 5
--max-request 2000
```

this would result in the request being made with 
```
body: {
  ...
  messages: [
    {role: "system", content: "Please be a helpful ai assistant that is funny and tells bad jokes."},
    {role: "user", content: "lets count one number at a time ill start 1"}
    {role: "assistant", content: "2"}
    {role: "user", content: "3"}
    {role: "assistant", content: "4"}
    {role: "user", content: "5, please summarize what we have done."}
  ]
}

...
```


Here is an example illustrating the problem with the lack of context.

```
> whats tha capital of america
As an AI language model, I must inform you that the capital of America is not a specific location as America refers to a continent and not a country. However, the capital of the United States of America is Washington, D.C.
Tokens: 53 Cost: $0.00011 Elapsed: 2.755 Seconds. Tokens and cost are estimates when streaming.
> whats the best thing to do there
As an AI language model, I don't know what you're referring to as "there" but here are some suggestions on what to do in general:
1. Spend time with loved ones - spend quality time with your family and friends or call them if you can't be with them in person.
2. Try a new hobby - pick up a new hobby that you've always wanted to try, such as painting, cooking or dance.
3. Exercise - get active and incorporate exercise into your daily routine to improve physical and mental health.
4. Learn something new - attend a workshop or take an online course to learn something new and stimulating.
5. Volunteer - help others and give back to your community by volunteering at a local organization or charity.
6. Travel - explore new places and cultures by traveling to a different city or country.
7. Relax and take some "me time"- take a break from your busy schedule and relax by reading, meditating or taking a relaxing bubble bath.
``` 


Hope this makes sense.  I think the psudo code is close, needs to update the interactive request execution or somthing. The system feature would involve extra though.

**Comment by drorm** [comment]
Ah yes, It's on my roadmap. I was thinking of it as "chat mode". But pretty much the same idea.

I'll probably generalize it so that it'll take arguments like to a printer where you can specify either a range or specific pages to print. Need to think a bit about what the right syntax to support that.

**Comment by drorm** [comment]
On further thought, it feels like the UI for this could get complicated quite quickly.
What do you think of, instead to:
1. Provide a way to save the history to a file,
2. Use a file as the source to a conversation.
So say you say gish -h 5, gish creates a file (using your example:
```
 messages: [
    {role: "user", content: "lets count one number at a time ill start 1"}
    {role: "assistant", content: "2"}
    {role: "user", content: "3"}
    {role: "assistant", content: "4"}
    {role: "user", content: "5, please summarize what we have done."}
  ]
```
You edit it to your heart's content, and when you exit, gish submits it.

Also, as far as summaries, my inclination is to summarize when needed rather than summarize everything so that we don't waste money on time unless it's needed. So for long content, there could be a summarizing step prior to the editing..

**Comment by syonfox** [comment]
json would be the most useful for a simple lib. then the end developer would be able to import in their lang of choice.  In the editor mode that would be cool if it just happened a new blank message.

Ie on file save it submits and adds one message and the empt space for a new one.. 

This is where message trimming could be useful. but it could keep extending the file and only submit the last n_responses or max_request_tokens.

**Comment by syonfox** [comment]
Yeah, it is all an optimization problem at the end of the day.  I think there is some threshold LEN such that it is cost-effective to summarize.  I think the natural solution will be to get gpt4 access and just submit large texts to that to summarize into an ideal input for 3.5-turbo.
